File: common/src/main/java/io/github/notenoughupdates/moulconfig/gui/CloseEventListener.java
package io.github.notenoughupdates.moulconfig.gui;

/**
 * Implement this interface alongside {@link GuiComponent} to receive an event when the gui closes.
 */
public interface CloseEventListener {
    enum CloseAction {
        NO_OBJECTIONS_TO_CLOSE,
        DENY_CLOSE;

        public CloseAction or(CloseAction other) {
            if (this == DENY_CLOSE) {
                return this;
            }
            return other;
        }
    }

    /**
     * Called just before a voluntary close. Return {@link CloseAction#DENY_CLOSE} to override the close. Make sure to update
     * your state such that the user can close the gui afterward.
     */
    default CloseAction onBeforeClose() {
        return CloseAction.NO_OBJECTIONS_TO_CLOSE;
    }

    /**
     * Called after the gui has been closed, both by the component gui, or by other actors, such as a teleport packet.
     */
    default void onAfterClose() {
    }
}

File: common/src/main/java/io/github/moulberry/moulconfig/gui/CloseEventListener.kt
package io.github.moulberry.moulconfig.gui

/**
 * Implement this interface alongside [GuiComponent] to receive an event when the gui closes.
 */
interface CloseEventListener {
    enum class CloseAction {
        NO_OBJECTIONS_TO_CLOSE, DENY_CLOSE, ;

        fun or(other: CloseAction): CloseAction {
            if (this == DENY_CLOSE) return this
            return other
        }
    }

    /**
     * Called just before a voluntary close. Return [CloseAction.DENY_CLOSE] to override the close. Make sure to update
     * your state such that the user can close the gui afterward.
     */
    fun onBeforeClose(): CloseAction

    /**
     * Called after the gui has been closed, both by the component gui, or by other actors, such as a teleport packet.
     */
    fun onAfterClose()
}